File: FlexTargetAndroid/app/src/main/java/com/flextarget/android/ui/ble/ConnectSmartTargetView.kt
package com.flextarget.android.ui.ble

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.layout.ContentScale
import coil.compose.AsyncImage
import com.flextarget.android.data.ble.BLEManager
import com.flextarget.android.data.ble.DiscoveredPeripheral
import com.flextarget.android.ui.imagecrop.ImageCropViewV2
import kotlinx.coroutines.delay

@Composable
fun ConnectSmartTargetView(
    bleManager: BLEManager = BLEManager.shared,
    onDismiss: () -> Unit,
    targetPeripheralName: String? = null,
    isAlreadyConnected: Boolean = false,
    onConnected: (() -> Unit)? = null
) {
    var statusText by remember { mutableStateOf("CONNECTING") }
    var showReconnect by remember { mutableStateOf(false) }
    var showProgress by remember { mutableStateOf(false) }
    var selectedPeripheral by remember { mutableStateOf<DiscoveredPeripheral?>(null) }
    var activeTargetName by remember { mutableStateOf<String?>(null) }
    var showImageCrop by remember { mutableStateOf(false) }
    var showBLETestTool by remember { mutableStateOf(false) }

    fun goToMain() {
        onConnected?.invoke() ?: onDismiss()
    }

    fun handleReconnect() {
        statusText = "Trying to connect..."
        showReconnect = false
        selectedPeripheral = null
        bleManager.startScan()
        showProgress = true
    }

    fun connectToSelectedPeripheral(peripheral: DiscoveredPeripheral) {
        selectedPeripheral = peripheral
        statusText = "Trying to connect..."
        showProgress = true

        bleManager.connectToSelectedPeripheral(peripheral)
    }

    // Handle initial state
    LaunchedEffect(Unit) {
        if (isAlreadyConnected) {
            statusText = "Target Connected"
        } else {
            // If a target peripheral name was passed in, begin scanning
            if (targetPeripheralName != null) {
                activeTargetName = targetPeripheralName
                statusText = "Trying to connect..."
                showReconnect = false
                selectedPeripheral = null
                bleManager.startScan()
                showProgress = true
                statusText = "Scanning for $targetPeripheralName"
            } else {
                statusText = "Ready to scan"
                showProgress = false
            }
        }
    }

    var hasHandledInitialConnection by remember { mutableStateOf(false) }

    // Handle connection state changes
    LaunchedEffect(bleManager.isConnected) {
        if (bleManager.isConnected && !hasHandledInitialConnection) {
            hasHandledInitialConnection = true
            if (!isAlreadyConnected) {
                statusText = "Connected"
                showReconnect = false
                showProgress = false
                goToMain()
            }
        }
    }

    // Handle scanning logic
    LaunchedEffect(bleManager.isScanning, activeTargetName) {
        if (bleManager.isScanning && activeTargetName != null) {
            delay(2000) // 2 second delay to allow BLE to power on
            if (bleManager.isScanning) {
                val target = activeTargetName
                if (target != null) {
                    val match = bleManager.discoveredPeripherals.find { it.name == target }
                    if (match != null) {
                        bleManager.stopScan()
                        connectToSelectedPeripheral(match)
                    } else {
                        // Target not found
                        bleManager.stopScan()
                        statusText = "Target not found"
                        showReconnect = true
                        showProgress = false
                    }
                } else if (bleManager.discoveredPeripherals.isEmpty()) {
                    bleManager.stopScan()
                    statusText = "No targets found"
                    showReconnect = true
                    showProgress = false
                }
            }
        }
    }

    // Handle connection timeout
    LaunchedEffect(selectedPeripheral) {
        selectedPeripheral?.let { peripheral ->
            delay(10000)
            if (!bleManager.isConnected && selectedPeripheral == peripheral) {
                bleManager.disconnect()
                statusText = "Bluetooth service not found"
                showReconnect = true
                showProgress = false
            }
        }
    }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black)
    ) {
        val configuration = LocalConfiguration.current
        val iconHeight = configuration.screenHeightDp.dp / 3

        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(16.dp)
                .verticalScroll(rememberScrollState()),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            // Visual Target Frame (using SVG icon)
            AsyncImage(
                model = "file:///android_asset/smart-target-icon.svg",
                contentDescription = "Smart Target Icon",
                modifier = Modifier
                    .height(iconHeight)
                    .padding(top = 20.dp),
                contentScale = ContentScale.Fit
            )

            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 16.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Status text with progress indicator
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        text = statusText,
                        color = Color.White,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Medium,
                        textAlign = TextAlign.Center
                    )

                    if (showProgress) {
                        CircularProgressIndicator(
                            modifier = Modifier.size(16.dp),
                            color = Color.White,
                            strokeWidth = 2.dp
                        )
                    }
                }

                // Reconnect button
                if (showReconnect) {
                    Button(
                        onClick = { handleReconnect() },
                        modifier = Modifier
                            .fillMaxWidth(0.75f)
                            .height(44.dp),
                        colors = ButtonDefaults.buttonColors(containerColor = Color.Red),
                        shape = RoundedCornerShape(8.dp)
                    ) {
                        Text(
                            text = "Reconnect",
                            color = Color.White,
                            fontSize = 20.sp,
                            fontWeight = FontWeight.Medium
                        )
                    }
                }

                // Connected state buttons
                if (isAlreadyConnected) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        modifier = Modifier.fillMaxWidth(),
                        verticalArrangement = Arrangement.spacedBy(16.dp)
                    ) {
                        Button(
                            onClick = {
                                bleManager.disconnect()
                                onDismiss()
                            },
                            modifier = Modifier
                                .fillMaxWidth(0.75f)
                                .height(44.dp),
                            colors = ButtonDefaults.buttonColors(containerColor = Color.Red),
                            shape = RoundedCornerShape(8.dp)
                        ) {
                            Text(
                                text = "Disconnect",
                                color = Color.White,
                                fontSize = 20.sp,
                                fontWeight = FontWeight.Medium
                            )
                        }

                        // Image Transfer button
                        Button(
                            onClick = {
                                showImageCrop = true
                            },
                            modifier = Modifier
                                .fillMaxWidth(0.75f)
                                .height(44.dp),
                            colors = ButtonDefaults.buttonColors(containerColor = Color.Blue),
                            shape = RoundedCornerShape(8.dp)
                        ) {
                            Text(
                                text = "My Target",
                                color = Color.White,
                                fontSize = 20.sp,
                                fontWeight = FontWeight.Medium
                            )
                        }

                        // BLE Test Tool button
                        Button(
                            onClick = {
                                showBLETestTool = true
                            },
                            modifier = Modifier
                                .fillMaxWidth(0.75f)
                                .height(44.dp),
                            colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF, 0x99, 0x00)),
                            shape = RoundedCornerShape(8.dp)
                        ) {
                            Text(
                                text = "BLE Test Tool",
                                color = Color.White,
                                fontSize = 20.sp,
                                fontWeight = FontWeight.Medium
                            )
                        }
                    }
                }
            }
        }

        // Close button (top right)
        IconButton(
            onClick = onDismiss,
            modifier = Modifier
                .align(Alignment.TopEnd)
                .padding(20.dp)
                .size(44.dp)
                .background(Color.White.copy(alpha = 0.2f), CircleShape)
        ) {
            Text(
                text = "✕",
                color = Color.White,
                fontSize = 20.sp
            )
        }

        // Error alert
        bleManager.error?.let { error ->
            AlertDialog(
                onDismissRequest = { /* Handle dismiss */ },
                title = { Text("Error") },
                text = { Text(error.message ?: "Unknown error occurred") },
                confirmButton = {
                    TextButton(onClick = { /* Handle OK */ }) {
                        Text("OK")
                    }
                }
            )
        }

        // Image Crop View
        if (showImageCrop) {
            ImageCropViewV2(
                onDismiss = { showImageCrop = false }
            )
        }

        // BLE Test Tool View
        if (showBLETestTool) {
            BLETestToolView(
                onDismiss = { showBLETestTool = false }
            )
        }
    }
}

File: FlexTargetiOS/flextarget/View/BLE/ConnectSmartTargetView.swift
import SwiftUI

struct ConnectSmartTargetView: View {
    @ObservedObject var bleManager: BLEManager
    @Binding var navigateToMain: Bool
    // Optional peripheral name passed in (from QR scan or navigation)
    var targetPeripheralName: String? = nil
    // Flag indicating if BLE is already connected
    var isAlreadyConnected: Bool = false
    var hideCloseButton: Bool = false
    @Environment(\.dismiss) var dismiss
    @State private var statusText: String = "CONNECTING"
    @State private var showReconnect: Bool = false
    @State private var isShaking: Bool = true
    @State private var showProgress: Bool = false
    @State private var hasTriedReconnect: Bool = false
    @State private var selectedPeripheral: DiscoveredPeripheral?
    @State private var showImageCrop: Bool = false
    @State private var connectionStartTime: Date?
    @State private var timeoutTimer: Timer?
    @State private var remainingSeconds: Int = 15
    @State private var scanStartTime: Date?
    @State private var minScanDurationTimer: Timer?
    private let minScanDuration: TimeInterval = 5.0  // Minimum 3 seconds before showing picker
    var onConnected: (() -> Void)?
    @State private var activeTargetName: String? = nil
    
    func goToMain() {
        if let onConnected = onConnected {
            onConnected()
        } else {
            dismiss()
        }
        navigateToMain = true
    }
    
    private struct TargetFrameView: View {
        let geometry: GeometryProxy
        let bleManager: BLEManager
        
        var body: some View {
            let frameWidth = geometry.size.width * 0.4
            let frameHeight = geometry.size.height * 0.35
            let dotPadding: CGFloat = 100
            let dotRadius: CGFloat = 6
            // Corner sensor icon configuration
            let sensorIconSize: CGFloat = 24
            let sensorOffsetAdjustment: CGFloat = -4 // how far outside the rectangle the icons sit
            let baseAnimation = Animation.easeInOut(duration: 0.6).repeatForever(autoreverses: true)
            let animation0 = baseAnimation
            let animation15 = baseAnimation.delay(0.15)
            let animation30 = baseAnimation.delay(0.30)
            let animation45 = baseAnimation.delay(0.45)
            
            ZStack(alignment: .topLeading) {
                Rectangle()
                    .stroke(Color.white, lineWidth: 10)
                    .frame(width: frameWidth, height: frameHeight)
                Circle()
                    .fill(Color(red: 0.8705882352941177, green: 0.2196078431372549, blue: 0.13725490196078433))
                    .frame(width: dotRadius * 2, height: dotRadius * 2)
                    .offset(x: dotPadding, y: dotPadding)
                
                // Bottom-left sensor icon (45° clockwise)
                Image(systemName: "dot.radiowaves.forward")
                    .font(.system(size: sensorIconSize))
                    .foregroundColor(.white)
                    .scaleEffect(!bleManager.isConnected ? 1.1 : 0.9)
                    .opacity(!bleManager.isConnected ? 1.0 : 0.6)
                    .animation(animation0, value: bleManager.isConnected)
                    .rotationEffect(.degrees(-45)) // clockwise
                    .offset(x: -sensorOffsetAdjustment, y: frameHeight - sensorIconSize + sensorOffsetAdjustment)
                
                // Bottom-right sensor icon (135° clockwise)
                Image(systemName: "dot.radiowaves.forward")
                    .font(.system(size: sensorIconSize))
                    .foregroundColor(.white)
                    .scaleEffect(!bleManager.isConnected ? 1.1 : 0.9)
                    .opacity(!bleManager.isConnected ? 1.0 : 0.6)
                    .animation(animation15, value: bleManager.isConnected)
                    .rotationEffect(.degrees(-135)) // clockwise
                    .offset(x: frameWidth - sensorIconSize + sensorOffsetAdjustment, y: frameHeight - sensorIconSize + sensorOffsetAdjustment)
                
                // Top-right sensor icon (135° counter-clockwise)
                Image(systemName: "dot.radiowaves.forward")
                    .font(.system(size: sensorIconSize))
                    .foregroundColor(.white)
                    .scaleEffect(!bleManager.isConnected ? 1.1 : 0.9)
                    .opacity(!bleManager.isConnected ? 1.0 : 0.6)
                    .animation(animation30, value: bleManager.isConnected)
                    .rotationEffect(.degrees(135)) // counter-clockwise
                    .offset(x: frameWidth - sensorIconSize + sensorOffsetAdjustment, y: -sensorOffsetAdjustment)
                
                // Top-left sensor icon (45° counter-clockwise)
                Image(systemName: "dot.radiowaves.forward")
                    .font(.system(size: sensorIconSize))
                    .foregroundColor(.white)
                    .scaleEffect(!bleManager.isConnected ? 1.1 : 0.9)
                    .opacity(!bleManager.isConnected ? 1.0 : 0.6)
                    .animation(animation45, value: bleManager.isConnected)
                    .rotationEffect(.degrees(45)) // counter-clockwise
                    .offset(x: -sensorOffsetAdjustment, y: -sensorOffsetAdjustment)
            }
            .padding(.top, geometry.size.height * 0.15)
        }
    }
    
    private struct StatusAndButtonsView: View {
        let geometry: GeometryProxy
        let statusText: String
        let showProgress: Bool
        let showReconnect: Bool
        let isAlreadyConnected: Bool
        let bleManager: BLEManager
        let handleReconnect: () -> Void
        @Environment(\.dismiss) var dismiss
        
        var body: some View {
            VStack(spacing: 12) {
                HStack(spacing: 8) {
                    Text(statusText)
                        .font(.custom("SFPro-Medium", size: 16))
                        .foregroundColor(.white)
                        .multilineTextAlignment(.center)
                    if showProgress {
                        ProgressView()
                            .progressViewStyle(CircularProgressViewStyle(tint: .white))
                    }
                }
                .padding(.horizontal, 24)
                .padding(.vertical, 16)
                
                if showReconnect {
                    Button(action: handleReconnect) {
                        Text(NSLocalizedString("reconnect", comment: "Reconnect button"))
                            .font(.custom("SFPro-Medium", size: 20))
                            .foregroundColor(.white)
                            .frame(width: geometry.size.width * 0.75, height: 44)
                            .background(Color(red: 0.8705882352941177, green: 0.2196078431372549, blue: 0.13725490196078433))
                            .cornerRadius(8)
                    }
                    .padding(.horizontal)
                }
                
                if isAlreadyConnected {
                    VStack(spacing: 12) {
                        Button(action: {
                            bleManager.disconnect()
                            dismiss()
                        }) {
                            Text(NSLocalizedString("disconnect", comment: "Disconnect button"))
                                .font(.custom("SFPro-Medium", size: 20))
                                .foregroundColor(.white)
                                .frame(height: 44)
                                .frame(maxWidth: .infinity)
                                .background(Color(red: 0.8705882352941177, green: 0.2196078431372549, blue: 0.13725490196078433))
                                .cornerRadius(8)
                        }
                        .padding(.horizontal, 16)
                    }
                }
            }
            .frame(maxWidth: .infinity)
        }
    }
    
    var body: some View {
        GeometryReader { geometry in
            VStack(spacing: 0) {
                TargetFrameView(geometry: geometry, bleManager: bleManager)
                
                StatusAndButtonsView(
                    geometry: geometry,
                    statusText: statusText,
                    showProgress: showProgress,
                    showReconnect: showReconnect,
                    isAlreadyConnected: isAlreadyConnected,
                    bleManager: bleManager,
                    handleReconnect: handleReconnect
                )
                .padding(.top, 120)
            }//Top Level VStack
            .frame(width: geometry.size.width, height: geometry.size.height, alignment: .top)
            //                .border(Color.white, width: 1)
        }//Top Level Geometry Reader
        .overlay(alignment: .topTrailing) {
            if !hideCloseButton {
                Button(action: { dismiss() }) {
                    Image(systemName: "xmark")
                        .foregroundColor(.white)
                        .font(.system(size: 20))
                        .padding(12)
                        .background(Color.white.opacity(0.2))
                        .clipShape(Circle())
                }
                .padding(.trailing, 20)
                .padding(.top, 20)
                .accessibilityLabel(Text("Close"))
            }
        }
        .sheet(isPresented: $showImageCrop) {
            ImageCropView()
        }
        .sheet(isPresented: $bleManager.showMultiDevicePicker) {
            MultiDevicePickerSheetView(bleManager: bleManager)
        }
        .background(Color.black.ignoresSafeArea())
        .alert(isPresented: $bleManager.showErrorAlert) {
            Alert(
                title: Text("Error"),
                message: Text(bleManager.errorMessage ?? "Unknown error occurred"),
                dismissButton: .default(Text("OK"))
            )
        }
        .onChange(of: bleManager.error) { error in
            if case .bluetoothOff = error {
                statusText = "Bluetooth has been turned off"
                isShaking = false
                showProgress = false
                showReconnect = true
                timeoutTimer?.invalidate()
                
                // Dismiss after 2 seconds if already connected
                if isAlreadyConnected {
                    DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                        dismiss()
                    }
                }
            }
        }
        .onChange(of: bleManager.isConnected) { isConnected in
            if !isConnected && hasTriedReconnect && activeTargetName == nil {
                // Unexpected disconnection during auto-detect
                statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
                showReconnect = true
                showProgress = false
                isShaking = false
                timeoutTimer?.invalidate()
            }
        }
        .onAppear {
            if isAlreadyConnected {
                let deviceName = bleManager.connectedPeripheral?.name ?? "Device"
                let connectedText = NSLocalizedString("target_connected", comment: "Status when target is connected")
                statusText = deviceName + " " + connectedText
            } else {
                connectionStartTime = Date()
                startConnectionTimeout()
                // If a target peripheral name was passed in (from QR scan), begin scanning
                if let target = targetPeripheralName {
                    activeTargetName = target
                    bleManager.setAutoConnectTarget(target)
                    startScanAndTimer()
                    statusText = String(format: NSLocalizedString("scanning_for_target", comment: "Scanning for specific target"), target)
                } else if bleManager.autoDetectMode {
                    // Auto-detect mode: check if peripherals are already discovered
                    if bleManager.discoveredPeripherals.count == 1 {
                        // Single device found - auto-connect
                        let peripheral = bleManager.discoveredPeripherals.first!
                        bleManager.completeScan()
                        selectedPeripheral = peripheral
                        showProgress = true
                        statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
                        connectToSelectedPeripheral()
                    } else if bleManager.discoveredPeripherals.count > 1 {
                        // Multiple devices found - show selection
                        bleManager.stopScan()
                        bleManager.showMultiDevicePicker = true
                        statusText = "Multiple devices found, select one"
                        showProgress = false
                    } else if bleManager.isScanning {
                        // Scanning in progress, wait for results
                        statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
                        showProgress = true
                        isShaking = true
                    } else {
                        // No devices found yet, start scanning if not already scanning
                        if !bleManager.isScanning {
                            bleManager.startScan()
                        }
                        statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
                        showProgress = true
                        isShaking = true
                    }
                } else {
                    // No target provided and not in auto-detect mode
                    statusText = NSLocalizedString("ready_to_scan", comment: "Ready to scan prompt")
                    showProgress = false
                    isShaking = false
                }
            }
        }
        .onDisappear {
            timeoutTimer?.invalidate()
            minScanDurationTimer?.invalidate()
            // Clear any auto-connect target on exit
            bleManager.setAutoConnectTarget(nil as String?)
        }
        .onChange(of: bleManager.isConnected) { newValue in
            if newValue {
                timeoutTimer?.invalidate()
                let deviceName = bleManager.connectedPeripheral?.name ?? "Device"
                let connectedText = NSLocalizedString("connected", comment: "Status when connection successful")
                statusText = deviceName + " " + connectedText
                isShaking = false
                showReconnect = false
                showProgress = false
                goToMain()
            }
        }
        .onChange(of: bleManager.discoveredPeripherals) { newValue in
            // If we are looking for a specific target name, try to match and auto-connect
            if let target = activeTargetName, bleManager.isScanning {
                if let match = bleManager.findPeripheral(named: target) {
                    // Found the target peripheral - connect to it
                    bleManager.completeScan()
                    selectedPeripheral = match
                    showProgress = true
                    connectToSelectedPeripheral()
                    return
                }
                // Otherwise wait for timers (min duration / overall timeout) to decide
                return
            }
            // Auto-detect mode
            if bleManager.autoDetectMode && activeTargetName == nil {
                if bleManager.discoveredPeripherals.count == 1 {
                    let peripheral = bleManager.discoveredPeripherals.first!
                    bleManager.completeScan()
                    selectedPeripheral = peripheral
                    showProgress = true
                    connectToSelectedPeripheral()
                } else if bleManager.discoveredPeripherals.count > 1 {
                    bleManager.stopScan()
                    bleManager.showMultiDevicePicker = true
                    statusText = "Multiple devices found, select one"
                    showProgress = false
                }
            }
        }
    }
    
    func startScanAndTimer() {
        statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
        isShaking = true
        showReconnect = false
        selectedPeripheral = nil
        scanStartTime = Date()  // Record scan start time
        bleManager.startScan()
        showProgress = true
        minScanDurationTimer?.invalidate()
        
        // Delay the scan timer start to allow BLE to power on
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
            if self.bleManager.isScanning {
                // If a target name was requested, check for it and auto-connect or dismiss
                if let target = self.activeTargetName {
                    if let match = self.bleManager.findPeripheral(named: target) {
                        self.bleManager.completeScan()
                        self.selectedPeripheral = match
                        self.connectToSelectedPeripheral()
                    } else {
                        // Target not found in scan results -> dismiss
                        self.bleManager.completeScan()
                        // Clear manager auto-connect target before dismissing
                        self.bleManager.setAutoConnectTarget(nil as String?)
                        self.dismiss()
                    }
                } else {
                    if self.bleManager.discoveredPeripherals.isEmpty {
                        // Scan timeout with no peripherals found
                        self.bleManager.completeScan()
                        self.statusText = NSLocalizedString("target_not_found", comment: "Status when no targets found after scan")
                        self.isShaking = false
                        self.showReconnect = true
                        self.showProgress = false
                    }
                }
            }
        }
    }
        
        func handleReconnect() {
            hasTriedReconnect = true
            startScanAndTimer()
        }
        
        func connectToSelectedPeripheral() {
            guard let peripheral = selectedPeripheral else {
                return
            }
            statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device")
            showProgress = true
            
            bleManager.connectToSelectedPeripheral(peripheral)
            // Clear manager auto-connect target since we are actively connecting
            bleManager.setAutoConnectTarget(nil as String?)
            
            // Start 30s connection timer (increased from 10s to match BLEManager timeout)
            DispatchQueue.main.asyncAfter(deadline: .now() + 30) {
                if !self.bleManager.isConnected {
                    // Connection timeout
                    self.bleManager.disconnect()
                    self.statusText = NSLocalizedString("bluetooth_service_not_found", comment: "Status when bluetooth service not found during connection")
                    self.isShaking = false
                    self.showReconnect = true
                    self.showProgress = false
                }
            }
        }
        
        func startConnectionTimeout() {
            timeoutTimer?.invalidate()
            remainingSeconds = 45  // Increased from 15 to 45 seconds
            
            timeoutTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { _ in
                guard let startTime = connectionStartTime else { return }
                let elapsedSeconds = Int(Date().timeIntervalSince(startTime))
                remainingSeconds = max(0, 45 - elapsedSeconds)  // Updated to match new timeout
                
                // Update status text with countdown
                statusText = NSLocalizedString("trying_to_connect", comment: "Status when trying to connect to FlexTarget device") + " (\(remainingSeconds))"
                
                // If 45 seconds have passed and not connected, dismiss
                if elapsedSeconds >= 45 && !bleManager.isConnected {
                    timeoutTimer?.invalidate()
                    bleManager.disconnect()
                    bleManager.completeScan()
                    dismiss()
                }
            }
        }
    }
    
    // MARK: - Preview
#if DEBUG
    struct ConnectSmartTargetView_Previews: PreviewProvider {
        static var previews: some View {
            // Provide a constant binding for navigateToMain. Use shared BLEManager for preview.
            ConnectSmartTargetView(bleManager: BLEManager.shared, navigateToMain: .constant(false), isAlreadyConnected: false)
                .previewLayout(.fixed(width: 375, height: 700))
                .background(Color.black)
        }
    }
#endif

// MARK: - Multi-Device Picker Sheet
struct MultiDevicePickerSheetView: View {
    var bleManager: BLEManager
    @Environment(\.dismiss) var dismiss
    
    var body: some View {
        ZStack {
            Color.black.ignoresSafeArea()
            
            VStack(spacing: 0) {
                // Header
                VStack(spacing: 12) {
                    Text("Select a Device")
                        .font(.custom("SFPro-Bold", size: 24))
                        .foregroundColor(.white)
                    
                    Text("Multiple targets found. Please select one to connect:")
                        .font(.custom("SFPro-Regular", size: 14))
                        .foregroundColor(.gray)
                        .multilineTextAlignment(.center)
                }
                .padding(.vertical, 24)
                .padding(.horizontal, 20)
                
                // Device List
                ScrollView {
                    VStack(spacing: 12) {
                        ForEach(bleManager.discoveredPeripherals) { peripheral in
                            DeviceSelectionButtonView(
                                peripheral: peripheral,
                                action: {
                                    bleManager.selectDeviceFromPicker(peripheral)
                                }
                            )
                        }
                    }
                    .padding(.horizontal, 20)
                    .padding(.vertical, 12)
                }
                
                // Cancel Button
                VStack(spacing: 0) {
                    Divider()
                        .background(Color.white.opacity(0.1))
                    
                    Button(action: {
                        bleManager.dismissDevicePicker()
                    }) {
                        Text("Cancel")
                            .font(.custom("SFPro-Medium", size: 18))
                            .foregroundColor(.white)
                            .frame(maxWidth: .infinity)
                            .frame(height: 50)
                    }
                    .padding(.horizontal, 20)
                    .padding(.vertical, 16)
                }
            }
        }
    }
}

struct DeviceSelectionButtonView: View {
    let peripheral: DiscoveredPeripheral
    let action: () -> Void
    @State private var isPressed = false
    
    var body: some View {
        Button(action: action) {
            HStack(spacing: 16) {
                // Device Icon
                Image(systemName: "smartphone.badge.checkmark")
                    .font(.system(size: 24))
                    .foregroundColor(Color(red: 0.8705882352941177, green: 0.2196078431372549, blue: 0.13725490196078433))
                
                // Device Information
                VStack(alignment: .leading, spacing: 4) {
                    Text(peripheral.name)
                        .font(.custom("SFPro-Medium", size: 16))
                        .foregroundColor(.white)
                        .lineLimit(1)
                    
                    Text(peripheral.peripheral.identifier.uuidString.prefix(12).uppercased())
                        .font(.custom("SFPro-Regular", size: 12))
                        .foregroundColor(.gray)
                }
                
                Spacer()
                
                // Arrow Icon
                Image(systemName: "chevron.right")
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(.gray)
            }
            .padding(.vertical, 14)
            .padding(.horizontal, 16)
            .background(Color.white.opacity(0.05))
            .cornerRadius(8)
            .overlay(
                RoundedRectangle(cornerRadius: 8)
                    .stroke(Color.white.opacity(0.1), lineWidth: 1)
            )
            .scaleEffect(isPressed ? 0.95 : 1.0)
            .onTapGesture {
                withAnimation(.easeInOut(duration: 0.1)) {
                    isPressed = true
                }
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    action()
                    withAnimation(.easeInOut(duration: 0.1)) {
                        isPressed = false
                    }
                }
            }
        }
    }
}